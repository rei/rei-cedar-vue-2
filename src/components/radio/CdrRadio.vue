<template>
  <div :class="$style['cdr-radio__wrap']">
    <label
      :class="[modifierClass, labelClass]"
      ref="label"
    >
      <input
        :class="[$style['cdr-radio__input'], inputClass]"
        type="radio"
        v-bind="$attrs"
        :name="name"
        :checked="isChecked"
        @change="onChange"
        :value="value"
        ref="radio"
      >
      <span :class="$style['cdr-radio__figure']" />
      <div :class="[$style['cdr-radio__content'], contentClass]">
        <!-- @slot innerHTML inside of radio component -->
        <slot />
      </div>
    </label>
  </div>
</template>

<script>
import modifier from 'mixinsdir/modifier';
import isEqual from 'lodash/isEqual';

/**
 * Cedar 2 component for radio
 * **NOTE:** `v-model` is required.
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrRadio',
  mixins: [modifier],
  inheritAttrs: false,
  model: {
    prop: 'modelValue',
    event: 'change',
  },
  props: {
    /**
     * Class that is added to the label for custom styles
     */
    labelClass: String,
    /**
     * Class that is added to the input for custom styles
     */
    inputClass: String,
    /**
     * Class that is added to the slot wrapper for custom styles
     */
    contentClass: String,
    /**
     * Sets the name of the radio. Required.
    */
    name: {
      type: String,
      required: true,
    },
    /**
     * Sets the value of the radio. Required.
    */
    value: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      required: true,
    },
    /** @ignore */
    modelValue: {
      type: [String, Number, Boolean, Object, Array, Symbol, Function],
      required: false,
    },
  },
  computed: {
    isChecked() {
      return isEqual(this.modelValue, this.value);
    },
    baseClass() {
      return 'cdr-radio';
    },
  },
  methods: {
    onChange() {
    /**
     * Selected radio value. Fires on section.
     * @event change
     * @type boolean|array
     */
      this.$emit('change', this.value);
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/CdrRadio.pcss';
</style>
