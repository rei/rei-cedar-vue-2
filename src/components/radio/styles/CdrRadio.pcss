/* ==========================================================================
  # CdrRadio
  
  All values should map to variables in
  vars/CdrRadio.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Radio
      :Elements
        :wrap
        :input
        :figure
      :States
      :Modifiers
        :compact
        :hide-figure
========================================================================== */

.cdr-radio {
  @include redwood-display-30;

  color: $radio-label-color;
  display: inline-block;
  margin: 0;
  position: relative;
  padding-left: calc($radio-figure-size + $radio-figure-space-right);

  /* Elements
    ========================================================================== */

  /* Wrap
  ========== */
  &__wrap {
    /* get rid of inline-block whitespace from label element */
    font-size: 0;
  }

  /* Input
  ========== */
  &__input {
    position: absolute;
    opacity: 0;
    z-index: -1;
  }

  /* Figure
  ========== */

  &__figure {
    position: absolute;
    top: $radio-figure-space-top;
    left: 0;
    width: $radio-figure-size;
    height: $radio-figure-size;
    border-radius: $radio-figure-radius;
    background-color: $radio-figure-background-color;
    border: 1px solid $radio-figure-border-color;
    
    &::after {
      content: '';
      position: absolute;
      width: $radio-figure-selected-size;
      height: $radio-figure-selected-size;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%);
      border-radius: $radio-figure-radius;
    }
  }

  /* States
    ========================================================================== */

  /* Checked
  ========== */
  &__input:checked ~ &__figure {
    border-color: $radio-figure-checked-border-color;
    
    &::after {
      background-color: $radio-figure-checked-background-color;
    }
  }

  /* Hover */
  &:hover {
    & > .cdr-radio__figure {
      cursor: pointer;
      border-color: $radio-figure-hover-border-color;
      background-color: $radio-figure-hover-background-color;
    }

    & > .cdr-radio__content {
      cursor: pointer;
    }

    & > .cdr-radio__input {
      /* Hover + Checked
      ========== */
      &:checked ~ .cdr-radio__figure {
        border-color: $radio-figure-checked-hover-border-color;

        &::after {
          background-color: $radio-figure-checked-hover-background-color;
        }
      }
    }
  }

  /* Disabled
  ========== */
  &__input:disabled {
    & ~ .cdr-radio__content {
      color: $radio-label-disabled-color !important;
      cursor: default;
    }

    & ~ .cdr-radio__figure {
      border-color: $radio-figure-disabled-border-color !important;
      background-color: $radio-figure-disabled-background-color !important;
      cursor: default;
    }

    /* Disabled + Checked
    ========== */
    &:checked {
      & ~ .cdr-radio__figure {
        &::after {
          background-color: $radio-figure-checked-disabled-background-color !important;
        }
      }
    }
  }

  /* Active
  ========== */
  &:active > .cdr-radio__figure {
    border-color: $radio-figure-active-border-color;
    background-color: $radio-figure-background-color;
  }

  /* Focus
  ========== */
  &__input:focus ~ .cdr-radio__figure {
    border-color: $radio-figure-focus-border-color;
    box-shadow: $radio-focus-outline;
  }

  /* Modifiers
    ========================================================================== */

  /* Compact
    ========== */

  &--compact {
    @include redwood-display-20;

    & > .cdr-radio__figure {
      top: $radio-compact-figure-space-top;
    }
  }

  /* Hide Figure
    ========== */

  &--hide-figure {
    padding-left: 0;

    & > .cdr-radio__figure {
      display: none;
    }
  }
}
