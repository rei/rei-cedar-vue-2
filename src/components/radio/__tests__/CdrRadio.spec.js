import { shallowMount, mount } from '@vue/test-utils';
import CdrRadio from 'componentdir/radio/CdrRadio';

describe('CdrRadio', () => {
  it('renders an input', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('is type radio', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
      }
    });
    expect(wrapper.vm.$refs.radio.hasAttribute('type', 'radio')).toBe(true);
  });

  it('adds a custom label class correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        labelClass: 'custom-label-class',
        customValue: 'A',
        name: 'testName',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.vm.$refs.label.classList.contains('custom-label-class')).toBe(true);
  });

  it('adds a custom input class correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        inputClass: 'custom-input-class',
        customValue: 'A',
        name: 'testName',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.vm.$refs.radio.classList.contains('custom-input-class')).toBe(true);
  });

  it('adds a custom content class correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        contentClass: 'custom-content-class',
        customValue: 'A',
        name: 'testName',
      },
      slots: {
        default: 'Label Test',
      },
    });
    expect(wrapper.find('.custom-content-class').exists()).toBe(true);
  });

  it('sets name attribute correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
      }
    });
    expect(wrapper.vm.$refs.radio.hasAttribute('name', 'testName')).toBe(true);
  });

  it('evaluates simple not checked state correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
        value: 'AA',
      },
    });
    expect(wrapper.vm.$refs.radio.checked).toBe(false);
  });

  it('evaluates simple checked state correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
        value: 'A',
      },
    });
    expect(wrapper.vm.$refs.radio.checked).toBe(true);
  });

  it('evaluates complex group not checked state correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: {test: 'B', arr: [1,2,3]},
        name: 'testName',
        value: {test: 'B'},
      },
    });
    expect(wrapper.vm.$refs.radio.checked).toBe(false);
  });

  it('evaluates complex group checked state correctly', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: {test: 'B', arr: [1,2,3]},
        name: 'testName',
        value: {test: 'B', arr: [1,2,3]},
      },
    });
    expect(wrapper.vm.$refs.radio.checked).toBe(true);
  });

  it('emits a change event with correct value', () => {
    const wrapper = shallowMount(CdrRadio, {
      propsData: {
        customValue: 'A',
        name: 'testName',
        value: '',
      },
    });

    const rb = wrapper.find({ ref: 'radio'});
    rb.trigger('click')
    expect(wrapper.emitted().change[0][0]).toBe('A');
  });
});
