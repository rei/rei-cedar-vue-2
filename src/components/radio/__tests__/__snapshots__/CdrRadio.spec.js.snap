// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrRadio renders an input 1`] = `
<div
  class="cdr-radio__wrap"
>
  <label
    class="cdr-radio"
  >
    <input
      class="cdr-radio__input"
      name="testName"
      type="radio"
    />
    <span
      class="cdr-radio__figure"
    />
    <div
      class="cdr-radio__content"
    />
  </label>
</div>
`;
